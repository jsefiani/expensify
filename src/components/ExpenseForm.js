import React from 'react';

// Package for displaying dates
import moment from 'moment';

// Import datepicker 
import { SingleDatePicker } from "react-dates";



class ExpenseForm extends React.Component {
    constructor(props){
        super(props);

        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: "",
            buttonText: props.buttonText
        }
    }

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({
            description
        }))
    }

    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({
            note
        }))
    }

    onAmountChange = (e) => {
        const amount = e.target.value;
        // Regular expression for only allowing numbers from 1 to infinity
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({
                amount
            }))
        }
    }

    onDateChange = (createdAt) => {
        if(createdAt) {
            this.setState(() => ({
                createdAt
            }))
        }
    }

    onFocusChange = ({ focused }) => {
        this.setState(() => ({
            calendarFocused: focused
        }));
    }

    onSubmit = (e) => {
        e.preventDefault();

        if(!this.state.description || !this.state.amount) {
            // Set error to "Please provide description and amount"
            this.setState(() => ({
                error: "Please provide a description and an amount"
            }));
        } else {
            this.setState(() => ({
                error: ""
            }));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(), // To get the unix timestamp
                note: this.state.note
            })
        }
    }
    
    render() {
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input 
                        type="text"
                        placeholder="Description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />
                    <input 
                        type="text"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker 
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />
                    <textarea
                        placeholder="Add a note for your expense (optional)"
                        value={this.state.note}
                        onChange={this.onNoteChange}
                    >
                    </textarea>
                    <button>{this.state.buttonText}</button>
                </form>
            </div>
        )
    }
}

export default ExpenseForm;