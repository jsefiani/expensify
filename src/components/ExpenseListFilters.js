import React from 'react';
import { connect } from 'react-redux';
// Import datepicker 
import { DateRangePicker} from "react-dates";

// Import actions
import { 
    setTextFilter, 
    sortByAmount, 
    sortByDate,
    setStartDate, 
    setEndDate 
} from "../actions/filtersActions";


export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null
    }

    onDatesChange = ({startDate, endDate}) => {
        this.props.setStartDate(startDate)
        this.props.setEndDate(endDate)
    }

    onFocusChange = (calendarFocused) => {
        this.setState(() => ({
            calendarFocused
        }))
    }

    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    }

    onFilterChange = (e) => {
        if (e.target.value === "date") {
            this.props.sortByDate()
        } else if (e.target.value === "amount") {
            this.props.sortByAmount();
        }
    }

    render() {
        return (
            <div>
                <input
                    value={this.props.filters.text}
                    onChange={this.onTextChange}
                    type="text"
                />
                <select
                    value={this.props.filters.sortBy}
                    onChange={this.onFilterChange}
                >
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    isOutsideRange={() => false} // Be able to choose past dates
                    numberOfMonths={1}
                    showClearDates={true} // Clear dates
                />
            </div>
        );
    }
}

const mapStateToProps = (state) => ({
    filters: state.filters
});

const mapDispatchToProps = (dispatch) => {
    return {
        setTextFilter: (text) => dispatch(setTextFilter(text)),
        sortByDate: () => dispatch(sortByDate()),
        sortByAmount: () => dispatch(sortByAmount()),
        setStartDate: (startDate) => dispatch(setStartDate(startDate)),
        setEndDate: (endDate) => dispatch(setEndDate(endDate))
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);

