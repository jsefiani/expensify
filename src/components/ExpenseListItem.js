import React from 'react';
import moment from 'moment';

import { Link } from 'react-router-dom';

const ExpenseListItem = ({ description, amount, createdAt, id}) => {
    return (
        <div>
            <Link to={`/edit/${id}`}><h3>{description}</h3></Link>
            <h4>{amount} - {moment(createdAt).format("MMMM Do, YYYY")}</h4>
        </div>
    );
}

export default ExpenseListItem;