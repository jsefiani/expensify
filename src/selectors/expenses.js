import moment from 'moment'

// Setting up filter function
// Get filtered expenses
const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter(expense => {
        const createdAtMoment = moment(expense.createdAt)
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        if (sortBy === 'date') {
            return a.createdAt < b.createdAt ? 1 : -1
        } else if (sortBy === 'amount') {
            // 1 means that "b" will come first in the list
            return a.amount < b.amount ? 1 : -1
        }
    });
}

export default getVisibleExpenses;