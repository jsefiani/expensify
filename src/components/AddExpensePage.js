import React from 'react';
import { connect } from 'react-redux';
import { addExpense } from '../actions/expensesActions'
import ExpenseForm from './ExpenseForm';

export class AddExpensePage extends React.Component {
    constructor(props) {
        super(props);
    }

    onSubmit = (expense) => {
        //Add returned expense object to the store 
        // props.dispatch(addExpense(expense))
        this.props.addExpense(expense);
        // Redirect user to dashboard
        this.props.history.push('/');
    }

    render() {
        return (
            <div>
                <h1>Add expense</h1>
                <ExpenseForm
                    buttonText="Add Expense"
                    onSubmit={this.onSubmit}
                />
            </div>
        );
    }
}


const mapDispatchToProps = (dispatch) => {
    return {
        addExpense: (expense) => dispatch(addExpense(expense))
    }
}

export default connect(undefined, mapDispatchToProps)(AddExpensePage);