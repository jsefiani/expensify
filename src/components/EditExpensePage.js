import React from 'react';
import {connect} from 'react-redux';

import ExpenseForm from '../components/ExpenseForm';

// Import actions
import { removeExpense, editExpense } from "../actions/expensesActions";

export class EditExpensePage extends React.Component {

    constructor(props) {
        super(props);
    }

    onSubmit = (expense) => {
        this.props.editExpense(this.props.expense.id, expense)
        this.props.history.push('/')
    }

    onRemoveExpense = () => {
        // this.props.dispatch(removeExpense({ id: this.props.expense.id }))
        this.props.removeExpense({id: this.props.expense.id})
        this.props.history.push('/')
    }

    render() {
        return (
            <div>
                <ExpenseForm
                    expense={this.props.expense}
                    buttonText="Save Expense"
                    onSubmit={this.onSubmit}
                />
                <button onClick={this.onRemoveExpense}>
                    Remove Expense
                </button>
            </div>
        );
    }
}

// Get access to the right expense 
const mapStateToProps = (state, props) => ({
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
})

const mapDispatchToProps = (dispatch, props) => {
    return {
        editExpense: (id, expense) => dispatch(editExpense(id, expense)),
        removeExpense: id => dispatch(removeExpense(id))
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);