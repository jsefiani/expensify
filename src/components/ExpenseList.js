import React from 'react';

// Importing component
import ExpenseListItem from './ExpenseListItem';

// Connect method for connecting component to Redux store
import { connect } from "react-redux";

// Importing filter function
import selectExpenses from '../selectors/expenses';

// Unconnected component
export const ExpenseList = (props) => (
    <div>
        {
            props.expenses.length === 0 ? (
                <p>No expenses found</p>
            ):(
                props.expenses.map(expense => {
                    return (
                        <ExpenseListItem key={expense.id} {...expense} />
                    )
                })
            )
        }
    </div>
);

// Getting data from store
const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    };
};

// Connecting component to Redux store
export default connect(mapStateToProps)(ExpenseList);
