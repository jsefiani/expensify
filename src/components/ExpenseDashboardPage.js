import React from 'react';

// Importing components
import ExpenseList from './ExpenseList';
import ExpenseListFilters from '../components/ExpenseListFilters';

const ExpenseDashboardPage = () => (
    <div>
        <ExpenseListFilters />
        <ExpenseList />
    </div>
);

export default ExpenseDashboardPage;