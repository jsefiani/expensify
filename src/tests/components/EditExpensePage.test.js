import React from 'react';
import { shallow } from "enzyme";
// Test data
import expenses from "../fixtures/expenses";
// Importing component
import { EditExpensePage } from "../../components/EditExpensePage";

let editExpense, removeExpense, wrapper, history;
beforeEach(() => {
    editExpense = jest.fn();
    removeExpense = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(<EditExpensePage
        history={history}
        editExpense={editExpense}
        removeExpense={removeExpense}
        expense={expenses[1]}
    />);
})

test('Should render EditExpensePage correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

test('Should handle editExpense', () => {
    wrapper.find("ExpenseForm").prop("onSubmit")(expenses[1]);
    // What should have happened?
    expect(history.push).toHaveBeenLastCalledWith("/");
    expect(editExpense).toHaveBeenLastCalledWith(expenses[1].id, expenses[1]); 
});

test('Should handle removeExpense', () => {
    wrapper.find("button").prop("onClick")();
    // What should have happened?
    expect(history.push).toHaveBeenCalledWith('/');
    expect(removeExpense).toHaveBeenCalledWith({id: expenses[1].id})
});