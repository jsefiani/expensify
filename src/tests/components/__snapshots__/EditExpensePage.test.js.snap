// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render EditExpensePage correctly 1`] = `
<div>
  <ExpenseForm
    buttonText="Save Expense"
    expense={
      Object {
        "amount": 109500,
        "createdAt": -345600000,
        "description": "Rent bills",
        "id": "2",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove Expense
  </button>
</div>
`;
